pwndbg/disasm/__init__.py

```diff
@@ -66,8 +72,54 @@
     "rv64": 22,
 }
 
+
+# Caching strategy:
+# To ensure we don't have stale register/memory information in our cached PwndbgInstruction,
+# we clear the cache whenever we DON'T do a `stepi`, `nexti`, `step`, or `next` command.
+# Although `stepi` and `nexti` always go to the next machine instruction in memory, `step` and `next`
+# can skip over multiple when GDB has debugging symbols and sourcecode
+# In order to determine that we did a `stepi`, `nexti`, `step`, or `next`, whenever the process stops,
+# we check if the current program counter is at the address of one of the instructions that we
+# emulated to the last time the process stopped. This allows use to skips a handful of instruction, but still retain the cache
+# Any larger changes of the program counter will cause the cache to reset.
+
+next_addresses_cache: set[int] = set()
+
+
+# Register GDB event listeners for all stop events
+@pwndbg.gdblib.events.stop
+def enhance_cache_listener() -> None:
+    # Clear the register value cache to ensure we get the correct program counter value
+    pwndbg.gdblib.regs.__getattr__.cache.clear()  # type: ignore[attr-defined]
+
+    if pwndbg.gdblib.regs.pc not in next_addresses_cache:
+        # Clear the enhanced instruction cache to ensure we don't use stale values
+        computed_instruction_cache.clear()
+
+
+@pwndbg.gdblib.events.mem_changed
+@pwndbg.gdblib.events.reg_changed
+def clear_on_reg_mem_change() -> None:
+
+    # We clear all the future computed instructions because when we manually change a register or memory, it's often a location
+    # used by the instructions at or just after the current PC, and our previously emulated future instructions might be inaccurate
+    computed_instruction_cache.pop(pwndbg.gdblib.regs.pc, None)
+
+    for addr in next_addresses_cache:
+        computed_instruction_cache.pop(addr, None)
+
+    next_addresses_cache.clear()
+
+
+# Dict of Address -> previous Address executed
+# This allows use to display the instructions that led to the current instruction
 backward_cache: DefaultDict[int, int] = collections.defaultdict(lambda: None)
 
+# This allows use to retain the annotation strings from previous instructions
```

Comment:
```suggestion
# This allows us to retain the annotation strings from previous instructions
```

Reply:
Nitpick: "annotation strings" can be thought as Python type annotations, heh. But w/e